util/network-devp2p/src/discovery.rs

```diff
@@ -541,10 +561,22 @@ impl<'a> Discovery<'a> {
 		};
 
 		if let Some((node, ping_reason)) = expected_node {
-			if let PingReason::FromDiscoveryRequest(target) = ping_reason {
+			if let PingReason::FromDiscoveryRequest(target, validity) = ping_reason {
 				self.respond_with_discovery(target, &node)?;
+				// kirushik: I would prefer to probe the network id of the remote node here, and add it to the nodes list if it's on "our" net --
+				// but `on_packet` happens synchronously, so doing the full TCP handshake ceremony here is a bad idea.
+				// So instead we just LRU-caching most recently seen nodes to avoid unnecessary pinging
+				match validity {
+					NodeValidity::UnknownNode | NodeValidity::ExpiredNode(NodeCategory::Observed) => {
+						self.other_observed_nodes.insert(node.id, (node.endpoint, Instant::now()));
+						Ok(None)
+					},
+					NodeValidity::ExpiredNode(NodeCategory::Bucket) => Ok(self.update_node(node)),
```

Comment:
I think we should call `update_node` but still return `Ok(None)` here.
We only want to update `node_buckets`, but not the peer table (i.e. we don't want `TableUpdates`).

Reply:
Note that this call may ping new nodes (it could cause amplification as well).
Should we pass the reason inside to avoid that?

Reply:
`update_node` doesn't ping the passed node directly — instead, it finds a least-recently-pinged with `select_bucket_ping` and pings that one. I don't think that this opens any space for targeted amplification.

The whole `update_node` logic can be refactored, though.